Add "Waiting For Goroutines To Finish" example.

eliben opened the pull request:
Also link to it from the Channel Synchronization example

mmcgrana commented on examples.txt:
```diff
@@ -23,6 +23,7 @@ Goroutines
 Channels
 Channel Buffering
 Channel Synchronization
+Waiting For Goroutines To Finish
```
If there's a word or small phrase in Go for the concept covered by an example I tend to use that, even if that word or phrase isn't intelligible on its own. E.g. "Select" instead of "Waiting on multiple channel operations". Sometimes there isn't a word in go, e.g. for "Non-Blocking Channel Operations". So here I'd favor title "WaitGroups" and slug `/waitgroups`.

mmcgrana replied:
I'm not sure where this example should go. This is one possibility. Another would be before or after the Worker Pool example (they seem related)? Indeed I'm curious to hear your thoughts on the question I linked from the corresponding issue another user filed about Worker Pools and WaitGroups.

eliben replied:
Alright, moving to `waitgroup`. Named it in singular to match the type, similarly to `Select` (not `Selects`), but if you feel that `WaitGroups` is better than `WaitGroup`, LMK. Also moving it to after the worker pool, which will mention it

mmcgrana replied:
Yes let's use the plural. Good observation on select, bu the slightly more nuanced convention is singlural for keyword syntax (if, else, switch, select) and plural for entities (maps, arrays, tickers, worker pools).

mmcgrana commented on examples/waiting-for-goroutines-to-finish/waiting-for-goroutines-to-finish.go:
```diff
@@ -0,0 +1,47 @@
+// To wait for multiple goroutines to finish, we can
+// use a sync.WaitGroup.
```
By convention we italic important terms when we first introduce them in an example.

mmcgrana replied:
We usually use the plain-English word(s) when introducing a concept and throughout the example when we refer to it in the text. For example in /timers we introduce "timers" and refer to the "timer" not `time.Timer`. WaitGroup is a little less obvious since it's not a standard CS phrase (I don't think?). But e.g. "we can use a _wait group_ ..." and "`wg` is the wait group it uses ..." doesn't seem obviously right either. I'm not sure.

eliben replied:
Changed heading comment, refactored the other comment.

mmcgrana commented on examples/waiting-for-goroutines-to-finish/waiting-for-goroutines-to-finish.go:
```diff
@@ -0,0 +1,47 @@
+)
+
+// This is the function we'll run in every goroutine.
+// wg is the WaitGroup it uses to notify that it's done.
```
variables form the code should be monotyped: "`wg` is the WaitGroup" etc.

eliben replied:
Removed this line since it's too obvious, IMHO and mentioned later too.

mmcgrana commented on examples/waiting-for-goroutines-to-finish/waiting-for-goroutines-to-finish.go:
```diff
@@ -0,0 +1,47 @@
+func worker(id int, wg *sync.WaitGroup) {
+    fmt.Printf("Worker %d starting\n", id)
+
+    // Sleep for a random duration between 500-700 ms
```
Do you feel like the randomization here is essential?

eliben replied:
Not essential, but for the presentation I think it's useful to not expect things to be deterministic. If all workers waited for time T, main could have just `time.Sleep(T+eps)` too wait for them, so the example is less convincing.

We could do something alternative like issue some HTTP HEAD requests to a few common URLs, instead.

I don't feel strongly about this, if you think we could replace the sleeps by:

```
// ... do work here
```

I'm OK with that too.

eliben replied:
I went ahead and changed that to just `time.Sleep(time.Second)` to match what the worker pool sample is doing. Consistency is important

mmcgrana commented:
@eliben the core of this example is solid; I think it'll prove a very helpful addition!

I've added some comments around presentation, relation to other examples, and relation to a good question in another issue. LMK what you think about those?

mmcgrana commented:
(I'm using assignee to indicate next action on the issue, so feel free to assign me after you respond. This just helps me differentiate which issues are on me to action vs. which I'm waiting on others for. But I don't feel strongly about this workflow, lmk if you have other preferences.)

eliben commented:
Thanks for the review, I'll do another round and will update the PR tomorrow.

Actually the assignee workflow works perfectly for me because I can check all the issues/PRs assigned to me easily instead of having to hunt for them. So feel free to assign to me issues where you ask a question and expect my answer too

eliben commented:
PTAL, I believe I've addressed the comments. There are some additional points to decide listed there (like the exact name). Also moved the sample to after worker pools and mentioned it there - seems like a natural location.

Fixes #166

mmcgrana commented on examples/channel-synchronization/channel-synchronization.go:
```diff
@@ -1,6 +1,8 @@
 // across goroutines. Here's an example of using a
 // blocking receive to wait for a goroutine to finish.
+// When waiting for multiple goroutines to finish,
+// you may prefer to [use a WaitGroup](waiting-for-goroutines-to-finish).
```
```suggestion
// you may prefer to use a [WaitGroup](waiting-for-goroutines-to-finish).
```

(as you do in the worker pools example.)

mmcgrana commented on examples/waitgroup/waitgroup.go:
```diff
@@ -0,0 +1,42 @@
+    time.Sleep(time.Second)
+    fmt.Printf("Worker %d done\n", id)
+
+    // Notify the WaitGroup that we're done.
```
```suggestion
    // Notify the WaitGroup that this worker is done.
```

maybe? just to make it clear that the _WaitGroup_ is not necessarily done.

mmcgrana commented on examples/waitgroup/waitgroup.go:
```diff
@@ -0,0 +1,42 @@
+    // goroutines launched here to finish.
+    var wg sync.WaitGroup
+
+    // Launch several goroutines and increment the WorkGroup
```
```suggestion
    // Launch several goroutines and increment the WaitGroup
```

I think? And one more below. Sorry I had seen this earlier but totally spaced on commenting on it.

mmcgrana commented:
Great! Some local comments. But otherwise I think this is ready! I'll take one more look once these final changes land but 🤞 your first example lands shortly!

eliben commented:
> Great! Some local comments. But otherwise I think this is ready! I'll take one more look once these final changes land but crossed_fingers your first example lands shortly!

I applied your suggestions via the Github interface - thanks!

mmcgrana commented on examples/waitgroup/waitgroup.go:
```diff
@@ -0,0 +1,42 @@
+        go worker(i, &wg)
+    }
+
+    // Block until the WorkGroup counter goes back to 0;
```
```suggestion
    // Block until the WaitGroup counter goes back to 0;
```

mmcgrana commented:
@eliben just flagging the pluralization comment above in case that got lost in the GitHub noise.

When you're ready please go ahead and assign back to me.

eliben commented:
> @eliben just flagging the pluralization comment above in case that got lost in the GitHub noise.
> 

Oops, yes. Will do

mmcgrana commented:
https://gobyexample.com/waitgroups 🎉
